pallets/payment/src/lib.rs

```diff
@@ -155,45 +155,48 @@ pub mod pallet {
 		/// This function will look for any pending scheduled tasks that can
 		/// be executed and will process them.
 		fn on_idle(now: T::BlockNumber, mut remaining_weight: Weight) -> Weight {
-			let mut task_list: Vec<(T::AccountId, T::AccountId, ScheduledTaskOf<T>)> =
-				ScheduledTasks::<T>::iter()
-					// leave out tasks in the future
-					.filter(|(_, _, ScheduledTask { when, .. })| when <= &now)
-					.collect();
-
-			if task_list.is_empty() {
-				return remaining_weight
-			} else {
-				task_list.sort_by(|(_, _, t), (_, _, x)| x.when.partial_cmp(&t.when).unwrap());
-			}
-
-			let cancel_weight =
-				T::WeightInfo::cancel().saturating_add(T::WeightInfo::remove_task());
-
-			while remaining_weight >= cancel_weight {
-				match task_list.pop() {
-					Some((from, to, ScheduledTask { task: Task::Cancel, .. })) => {
+			// reduce the weight used to read the task list
+			remaining_weight = remaining_weight.saturating_sub(T::WeightInfo::remove_task());
+
+			ScheduledTasks::<T>::mutate(|tasks| {
+				let tmp_tasks = tasks.clone();
```

Comment:
Don't think you need this clone since the `iter` doesn't mutate and `collect` creates a new vec

Reply:
The issue is this is an immutable borrow but tasks.remove later is a mutable borrow, let me see if it can be handled more elegantly

Reply:
🤔 the borrow is not needed after constructing the filtered list, maybe `{ scoping it in a block }` helps?

Reply:
I can't think of a way that looks good and makes the borrow checker happy 😅

Reply:
I tried that, the borrow checker does not like it, don't see a way to avoid the clone